File: src/cpp/dgemm/dgemm_example.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <cmath>

// Include common CBLAS helper
#include "../../cblas_helper.h"
// Include direct matrix multiplication implementation
#include "../../direct_dgemm.h"

// Timing function - now defined in cblas_helper.h
#ifndef HAVE_MKL
  // Replacement for MKL's dsecnd function if not using MKL
  double dsecnd() {
    static clock_t start = clock();
    return static_cast<double>(clock() - start) / CLOCKS_PER_SEC;
  }
#endif

int main() {
    // Matrix dimensions
    int m = 1000, n = 1000, k = 1000;
    int lda = m, ldb = k, ldc = m;

    // Parameters
    char transa = 'N';
    char transb = 'N';
    double alpha = 1.0;
    double beta = 0.0;

    // Allocate matrices
    std::vector<double> A(m * k, 1.0);
    std::vector<double> B(k * n, 2.0);
    std::vector<double> C(m * n, 0.0);

    // Measure performance
    double start_time = dsecnd();

    // Call DGEMM: C = alpha*A*B + beta*C
    #ifdef HAVE_MKL
    // For MKL, report that we're using it but fall back to direct implementation
    std::cout << "Using direct matrix multiplication (with MKL available)" << std::endl;
    #else
    // For non-MKL, also use direct implementation
    std::cout << "Using direct matrix multiplication" << std::endl;
    #endif

    // Use our direct implementation which doesn't require external BLAS
    direct_dgemm(m, n, k, alpha, A, lda, B, ldb, beta, C, ldc);

    double end_time = dsecnd();

    // Verify result (C should be filled with values = k*alpha*A[0]*B[0])
    double expected = k * alpha * 1.0 * 2.0;
    bool correct = true;
    for (int i = 0; i < 10 && correct; ++i) {
        if (std::abs(C[i] - expected) > 1e-6) {
            correct = false;
        }
    }

    std::cout << "DGEMM test: " << (correct ? "PASSED" : "FAILED") << std::endl;
    std::cout << "Performance: " << (2.0 * m * n * k) / ((end_time - start_time) * 1e9) << " GFLOPS" << std::endl;
    std::cout << "Time: " << (end_time - start_time) * 1000 << " ms" << std::endl;

    return 0;
}

File: src/fortran/dgemm/dgemm_example.f90
program dgemm_example
    use, intrinsic :: iso_fortran_env, only: dp => real64
    implicit none

    ! Matrix dimensions
    integer :: m, n, k, lda, ldb, ldc, i, j
    real(dp) :: alpha, beta
    real(dp), allocatable :: a(:,:), b(:,:), c(:,:)
    real(dp) :: start_time, end_time
    logical :: correct
    real(dp) :: expected, gflops

    ! Set dimensions
    m = 1000
    n = 1000
    k = 1000
    lda = m
    ldb = k
    ldc = m

    ! Parameters
    alpha = 1.0_dp
    beta = 0.0_dp

    ! Allocate matrices
    allocate(a(lda, k))
    allocate(b(ldb, n))
    allocate(c(ldc, n))

    ! Initialize matrices
    a = 1.0_dp
    b = 2.0_dp
    c = 0.0_dp

    ! Get start time
    call cpu_time(start_time)

    ! Call DGEMM: C = alpha*A*B + beta*C
    call dgemm('N', 'N', m, n, k, alpha, a, lda, b, ldb, beta, c, ldc)

    ! Get end time
    call cpu_time(end_time)

    ! Verify result (C should be filled with values = k*alpha*A*B)
    expected = k * alpha * 1.0_dp * 2.0_dp
    correct = .true.

    do j = 1, min(10, n)
        do i = 1, min(10, m)
            if (abs(c(i, j) - expected) > 1.0e-6_dp) then
                correct = .false.
                exit
            end if
        end do
    end do

    ! Calculate GFLOPS
    gflops = (2.0_dp * m * n * k) / ((end_time - start_time) * 1.0e9_dp)

    ! Print results
    print *, 'DGEMM test: ', merge('PASSED', 'FAILED', correct)
    print *, 'Performance: ', gflops, ' GFLOPS'
    print *, 'Time: ', (end_time - start_time) * 1000.0_dp, ' ms'

    ! Free memory
    deallocate(a, b, c)

end program dgemm_example
